moveit_ros/trajectory_cache/src/trajectory_cache.cpp

```diff
@@ -0,0 +1,1340 @@
+                                               const std::string& move_group_namespace,
+                                               const moveit_msgs::srv::GetCartesianPath::Request& plan_request,
+                                               const moveit_msgs::msg::RobotTrajectory& trajectory,
+                                               double execution_time_s, double planning_time_s, double fraction,
+                                               bool delete_worse_trajectories)
+{
+  // Check pre-conditions
+  if (!trajectory.multi_dof_joint_trajectory.points.empty())
+  {
+    RCLCPP_ERROR(node_->get_logger(), "Skipping cartesian trajectory insert: "
+                                      "Multi-DOF trajectory plans are not supported.");
+    return false;
+  }
+  if (plan_request.header.frame_id.empty() || trajectory.joint_trajectory.header.frame_id.empty())
+  {
+    RCLCPP_ERROR(node_->get_logger(), "Skipping cartesian trajectory insert: Frame IDs cannot be empty.");
+    return false;
+  }
+
+  auto coll = db_->openCollection<moveit_msgs::msg::RobotTrajectory>("move_group_cartesian_trajectory_cache",
+                                                                     move_group_namespace);
+
+  // Pull out trajectories "exactly" keyed by request in cache.
+  Query::Ptr exact_query = coll.createQuery();
+
+  bool start_query_ok =
+      this->extract_and_append_cartesian_trajectory_start_to_query(*exact_query, move_group, plan_request, 0);
+  bool goal_query_ok =
+      this->extract_and_append_cartesian_trajectory_goal_to_query(*exact_query, move_group, plan_request, 0);
+  exact_query->append("fraction", fraction);
+
+  if (!start_query_ok || !goal_query_ok)
+  {
+    RCLCPP_ERROR(node_->get_logger(), "Skipping cartesian trajectory insert: Could not construct lookup query.");
+    return false;
+  }
+
+  auto exact_matches = coll.queryList(exact_query, /* metadata_only */ true, /* sort_by */ "execution_time_s");
+  double best_execution_time = std::numeric_limits<double>::infinity();
+  if (!exact_matches.empty())
+  {
+    best_execution_time = exact_matches.at(0)->lookupDouble("execution_time_s");
+
+    if (delete_worse_trajectories)
+    {
+      for (auto& match : exact_matches)
+      {
+        double match_execution_time_s = match->lookupDouble("execution_time_s");
+        if (execution_time_s < match_execution_time_s)
+        {
+          int delete_id = match->lookupInt("id");
+          RCLCPP_INFO(node_->get_logger(),
+                      "Overwriting cartesian trajectory (id: %d): "
+                      "execution_time (%es) > new trajectory's execution_time (%es)",
+                      delete_id, match_execution_time_s, execution_time_s);
+
+          Query::Ptr delete_query = coll.createQuery();
+          delete_query->append("id", delete_id);
+          coll.removeMessages(delete_query);
+        }
+      }
+    }
+  }
+
+  // Insert if candidate is best seen.
```

Comment:
I guess best execution time might not always be the desired selection metric. Maybe sometimes users want the trajectory with the shortest path length or the least jerky trajectory. Would it be possible to expose the selection criteria as a function with the execution time being the default e.g. like here https://github.com/moveit/moveit2/pull/2153

Reply:
I'm not sure how we might execute this without significant tradeoffs in cache lookup.

The way the cache is handling what is "best" is by sorting by the execution time cache column.

If we want other things like path length or jerk, those values will need to be stored in the cache db (meaning increased space usage and incompatibility with older versions of the cache.) The most I can do for now I think is allow passing in a `sort_by` variable like I did with the fetching methods.

This would also mean needing to inject a function to extract those parameters and populate them in the cache as custom values, which I think is way out of the scope of this PR.

Reply:
RFE added: 241f6202102fc924c78b513ceaec26766b6ac6aa

Reply:
I am going to take a crack at doing this.

I am imagining something like passing in lambdas for "feature extractors" and "insert predicates" for the put and fetch methods.

It'll become clearer once I actually do it.

In the meantime it'll be super helpful to get a re-review so we don't have cascading changes while I'm trying to implement!

Reply:
The refactor is now complete (pending one test):
- https://github.com/moveit/moveit2/pull/2941

Example code is updated, builds, and runs as expected:
- https://github.com/moveit/moveit2_tutorials/pull/940